File: src/session.js
import { normalizeResponse } from "./normalizer.js";
import { sendEvent } from "./transport.js";
import { buildTrackUrl } from "./tracking.js";

function truncate(value, maxLen) {
  if (!value || !maxLen || typeof value !== "string") return value;
  if (value.length <= maxLen) return value;
  return value.slice(0, maxLen) + "...[truncated]";
}

export class Session {
  constructor(clientConfig, { sessionId, metadata } = {}) {
    Object.defineProperty(this, '_config', {
      value: clientConfig,
      enumerable: false,
    });
    this.sessionId = sessionId || crypto.randomUUID();
    this._metadata = metadata || {};
    this._reset();
  }

  _reset() {
    this._promptTokens = 0;
    this._completionTokens = 0;
    this._cacheReadTokens = 0;
    this._cacheCreateTokens = 0;
    this._rounds = 0;
    this._toolCalls = [];
    this._model = null;
    this._systemPrompt = null;
  }

  get totalUsage() {
    const usage = {
      prompt_tokens: this._promptTokens,
      completion_tokens: this._completionTokens,
      total_tokens:
        this._promptTokens +
        this._completionTokens +
        this._cacheReadTokens +
        this._cacheCreateTokens,
      ai_rounds: this._rounds,
    };
    // Cache token passthrough (Anthropic only). Added only when non-zero
    // so the legacy { prompt_tokens, completion_tokens, total_tokens,
    // ai_rounds } shape is preserved when no cache is in play. The
    // conversation-analytics Token Universe tab reads these directly.
    if (this._cacheReadTokens) {
      usage.cache_read_input_tokens = this._cacheReadTokens;
    }
    if (this._cacheCreateTokens) {
      usage.cache_creation_input_tokens = this._cacheCreateTokens;
    }
    return usage;
  }

  get toolCalls() {
    return this._toolCalls;
  }

  record(rawResponse) {
    const normalized = normalizeResponse(rawResponse);
    const round = this._rounds;

    this._promptTokens += normalized.usage.prompt_tokens;
    this._completionTokens += normalized.usage.completion_tokens;
    this._cacheReadTokens += normalized.usage.cache_read_input_tokens || 0;
    this._cacheCreateTokens += normalized.usage.cache_creation_input_tokens || 0;
    this._rounds += 1;

    if (normalized.model) {
      this._model = normalized.model;
    }

    for (const tc of normalized.toolCalls) {
      this._toolCalls.push({ ...tc, round });
    }

    return normalized;
  }

  /**
   * Attach a result summary to the most recent tool call matching `toolName`.
   * Call this after executing a tool to include results in the emitted event.
   */
  recordToolResult(toolName, result) {
    for (let i = this._toolCalls.length - 1; i >= 0; i--) {
      if (this._toolCalls[i].tool === toolName && !this._toolCalls[i].result) {
        this._toolCalls[i].result = result;
        return;
      }
    }
  }

  async emitChatTurn({ userMessage, assistantResponse, turnNumber, messages }) {
    const capture = this._config.captureContent !== false;
    const maxLen = this._config.maxContentLength;

    // Spread metadata first so SDK-controlled fields always win
    const attributes = {
      ...this._metadata,
      source: "pentatonic-ai-sdk",
      model: this._model,
      usage: this.totalUsage,
      tool_calls: this._toolCalls.length
        ? (capture ? this._toolCalls : this._toolCalls.map(({ args, ...rest }) => rest))
        : undefined,
    };

    if (capture) {
      attributes.user_message = truncate(userMessage, maxLen);
      attributes.assistant_response = truncate(assistantResponse, maxLen);
      if (this._systemPrompt) {
        attributes.system_prompt = truncate(this._systemPrompt, maxLen);
      }

      if (messages) {
        attributes.messages = messages.map((m) => {
          if (typeof m.content === "string") {
            return { ...m, content: truncate(m.content, maxLen) };
          }
          return m;
        });
      }
    }

    if (turnNumber !== undefined) {
      attributes.turn_number = turnNumber;
    }

    const result = await sendEvent(this._config, {
      eventType: "CHAT_TURN",
      entityType: "conversation",
      data: {
        entity_id: this.sessionId,
        attributes,
      },
    });

    this._reset();
    return result;
  }

  async emitToolUse({ tool, args, resultSummary, durationMs, turnNumber }) {
    const capture = this._config.captureContent !== false;
    const maxLen = this._config.maxContentLength;

    const attributes = {
      ...this._metadata,
      source: "pentatonic-ai-sdk",
      tool,
      duration_ms: durationMs,
      turn_number: turnNumber,
    };

    if (capture) {
      attributes.args = args;
      attributes.result_summary = truncate(resultSummary, maxLen);
    }

    // Spread metadata first so SDK-controlled fields always win
    return sendEvent(this._config, {
      eventType: "TOOL_USE",
      entityType: "conversation",
      data: {
        entity_id: this.sessionId,
        attributes,
      },
    });
  }

  async trackUrl(url, { eventType, attributes } = {}) {
    const payload = {
      u: url,
      s: this.sessionId,
      c: this._config.clientId,
      t: Math.floor(Date.now() / 1000),
      e: eventType || "LINK_CLICK",
    };
    const meta = { ...this._metadata, ...attributes };
    if (Object.keys(meta).length) {
      payload.a = meta;
    }
    return buildTrackUrl(this._config.endpoint, this._config.apiKey, payload);
  }

  async emitSessionStart() {
    return sendEvent(this._config, {
      eventType: "SESSION_START",
      entityType: "conversation",
      data: {
        entity_id: this.sessionId,
        attributes: {
          source: "pentatonic-ai-sdk",
          metadata: this._metadata,
        },
      },
    });
  }
}

File: python/pentatonic_agent_events/session.py
import uuid
from .normalizer import normalize_response
from .transport import send_event


def _truncate(value, max_len):
    if not value or not max_len or not isinstance(value, str):
        return value
    if len(value) <= max_len:
        return value
    return value[:max_len] + "...[truncated]"


class Session:
    def __init__(self, config, session_id=None, metadata=None):
        self._config = config
        self.session_id = session_id or str(uuid.uuid4())
        self._metadata = metadata or {}
        self._reset()

    def _reset(self):
        self._prompt_tokens = 0
        self._completion_tokens = 0
        self._rounds = 0
        self._tool_calls = []
        self._model = None

    @property
    def total_usage(self):
        return {
            "prompt_tokens": self._prompt_tokens,
            "completion_tokens": self._completion_tokens,
            "total_tokens": self._prompt_tokens + self._completion_tokens,
            "ai_rounds": self._rounds,
        }

    @property
    def tool_calls(self):
        return list(self._tool_calls)

    def record(self, raw_response):
        normalized = normalize_response(raw_response)
        current_round = self._rounds

        self._prompt_tokens += normalized["usage"]["prompt_tokens"]
        self._completion_tokens += normalized["usage"]["completion_tokens"]
        self._rounds += 1

        if normalized["model"]:
            self._model = normalized["model"]

        for tc in normalized["tool_calls"]:
            self._tool_calls.append({**tc, "round": current_round})

        return normalized

    def emit_chat_turn(self, user_message=None, assistant_response=None, turn_number=None, messages=None):
        capture = self._config.get("capture_content", True) is not False
        max_len = self._config.get("max_content_length")

        # Spread metadata first so SDK-controlled fields always win
        attributes = {
            **self._metadata,
            "source": "pentatonic-ai-sdk",
            "model": self._model,
            "usage": self.total_usage,
        }

        if self._tool_calls:
            if capture:
                attributes["tool_calls"] = self._tool_calls
            else:
                attributes["tool_calls"] = [
                    {k: v for k, v in tc.items() if k != "args"}
                    for tc in self._tool_calls
                ]

        if capture:
            attributes["user_message"] = _truncate(user_message, max_len)
            attributes["assistant_response"] = _truncate(assistant_response, max_len)

            if messages:
                attributes["messages"] = [
                    {**m, "content": _truncate(m["content"], max_len)}
                    if isinstance(m.get("content"), str) else m
                    for m in messages
                ]

        if turn_number is not None:
            attributes["turn_number"] = turn_number

        result = send_event(self._config, {
            "eventType": "CHAT_TURN",
            "entityType": "conversation",
            "data": {
                "entity_id": self.session_id,
                "attributes": attributes,
            },
        })

        self._reset()
        return result

    def emit_tool_use(self, tool=None, args=None, result_summary=None, duration_ms=None, turn_number=None):
        capture = self._config.get("capture_content", True) is not False
        max_len = self._config.get("max_content_length")

        attributes = {
            **self._metadata,
            "source": "pentatonic-ai-sdk",
            "tool": tool,
            "duration_ms": duration_ms,
            "turn_number": turn_number,
        }

        if capture:
            attributes["args"] = args
            attributes["result_summary"] = _truncate(result_summary, max_len) if isinstance(result_summary, str) else result_summary

        return send_event(self._config, {
            "eventType": "TOOL_USE",
            "entityType": "conversation",
            "data": {
                "entity_id": self.session_id,
                "attributes": attributes,
            },
        })

    def emit_session_start(self):
        return send_event(self._config, {
            "eventType": "SESSION_START",
            "entityType": "conversation",
            "data": {
                "entity_id": self.session_id,
                "attributes": {
                    "source": "pentatonic-ai-sdk",
                    "metadata": self._metadata,
                },
            },
        })
